src/index.rs

```diff
@@ -291,41 +296,95 @@ impl IndexRead {
+    /// - Every read index has its own unique read stats, therefore the clone does not inherit the read stats.
+    pub(crate) fn duplicate(&self) -> Self {
+        Self::open(self.transport.clone())
     }
 
-    /// Make an iterator that returns hunks of entries from this index.
-    pub fn iter_hunks(&self) -> IndexHunkIter {
-        let _span = debug_span!("iter_hunks", ?self.transport).entered();
-        // All hunk numbers present in all directories.
-        let subdirs = self
-            .transport
-            .list_dir("")
-            .expect("list index dir") // TODO: Don't panic
-            .dirs
-            .into_iter()
-            .sorted()
-            .collect_vec();
-        debug!(?subdirs);
+    /// Read and parse a specific hunk
+    pub fn read_hunk(&mut self, hunk_number: u32) -> Result<Option<Vec<IndexEntry>>> {
+        let path = hunk_relpath(hunk_number);
+        let compressed_bytes = match self.transport.read_file(&path) {
+            Ok(b) => b,
+            Err(err) if err.is_not_found() => {
+                // TODO: Cope with one hunk being missing, while there are still
+                // later-numbered hunks. This would require reading the whole
+                // list of hunks first.
+                return Ok(None);
+            }
+            Err(source) => return Err(Error::Transport { source }),
+        };
+        self.stats.index_hunks += 1;
+        self.stats.compressed_index_bytes += compressed_bytes.len() as u64;
+        let index_bytes = self.decompressor.decompress(&compressed_bytes)?;
+        self.stats.uncompressed_index_bytes += index_bytes.len() as u64;
+        let entries: Vec<IndexEntry> =
+            serde_json::from_slice(&index_bytes).map_err(|source| Error::DeserializeJson {
+                path: path.clone(),
+                source,
+            })?;
+        if entries.is_empty() {
+            // It's legal, it's just weird - and it can be produced by some old Conserve versions.
+        }
+        Ok(Some(entries))
+    }
+
+    // All hunk numbers present in all directories.
+    pub fn hunks_available(&self) -> Result<Vec<u32>> {
+        let subdirs = self.transport.list_dir("")?.dirs.into_iter().sorted();
+
         let hunks = subdirs
-            .into_iter()
             .filter_map(|dir| self.transport.list_dir(&dir).ok())
             .flat_map(|list| list.files)
             .filter_map(|f| f.parse::<u32>().ok())
             .sorted()
             .collect_vec();
+
+        Ok(hunks)
+    }
+
+    /// Make an iterator that will return all entries in this band.
+    pub fn iter_entries(self) -> IndexEntryIter<IndexHunkIter> {
+        // TODO: An option to pass in a subtree?
+        IndexEntryIter::new(
+            self.iter_available_hunks(),
+            Apath::root(),
+            Exclude::nothing(),
+        )
+    }
+
+    /// Make an iterator that returns hunks of entries from this index.
+    pub fn iter_available_hunks(self) -> IndexHunkIter {
```

Comment:
Is the change from `iter_hunks` to `iter_available_hunks` really semantic? They seem fairly similar..?

Reply:
Ahh this is actually a relict of the previous versions where I needed to pass a custom hunk iter to the `iter_hunks` method. Even tough, this is not functionally required any more I think the overall API semantics of IndexRead with `iter_hunks` and `iter_available_hunks` are improved.